usb_cdc: sam0: Add an option to keep ardurno-style upgrade enter

KubaFYI opened the pull request:
Many of the Zephyr supported boards are Arduino compatible.
However after being flashed with Zephyr, even if they keep their
original bootloader they lose the ability to have the firmware
uploaded to them hands-free, from the Arduino IDE. This commit
allows for enabling of such ability.

Currently only enabled for adafruit_feather_m0_basic_proto board
but the code should work in principle with other adafruit M0 SoCs.

Depends on #25361 for the flashing part to work.

mbolivar-nordic commented on subsys/power/reboot.c:
```diff
@@ -33,3 +36,45 @@ void sys_reboot(int type)
+	sys_clock_disable();
+#endif
+
+#ifdef CONFIG_SOC_SERIES_SAMD21
```
I like the goal of the PR, but I think this needs better abstraction / layering. I don't know this subsystem well enough to say exactly how it should go, though.

KubaFYI replied:
I agree that the abstraction here is not ideal but I don't really see a good alternative other than creating a totally new subsystem for enabling booting into a arduino-compatible samd bootloader which seems like an overkill.

I didn't want to allow much code to run between the erasure of the first line of code section and the reset and so I made the two take place in the same routine. And as it essentially creates a slightly different-flavoured kind or reset, I thought it would be appropriate it lives next to the vanilla reset code.

mbolivar-nordic replied:
To be honest, I very much doubt this PR will be merged without a better abstraction, but I'll remove my -1 for the Python bits.

mbolivar-nordic commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -58,6 +60,14 @@ def do_run(self, command, **kwargs):
                         'eof', '255']
             self.check_call(cmd_stty)
+            sleep(1)
+            while(True):
```
I think this should bail after a fixed number of retries, say 3, instead.

KubaFYI replied:
That's fair enough. Changed.

mbolivar-nordic reviewed (dismissed):
My -1 is for the bossac.py portion only.

KubaFYI commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -239,6 +243,13 @@ int cdc_acm_class_handle_req(struct usb_setup_packet *pSetup,
 		LOG_DBG("CDC_SET_CONTROL_LINE_STATE 0x%x",
 			dev_data->line_state);
+#ifdef CONFIG_ARDUINO_LIKE_UPGRADE
+		if(sys_le32_to_cpu(dev_data->line_coding.dwDTERate) == 1200 && 
```
> It also looks naive to me to reboot to bootloader based on baud rate setting.

I agree but that is that's mechanism those Backward-compatible boards use.

> I suggest to use non hand-free update procedure.

It's going to remain the default approach. I hid this code behind an `#ifdef` because I wanted there to be at least an _option_ to enable it and thus have the boards remain compatible with Arduino IDE should the user wish to do so.

> There should not be any platform/bootloader/update-technology code in the CDC ACM implementation.

Where do you suggest to put this then in an absence of a way to provide an external class_hande callback?

jfischer-phytec-iot replied:
> Where do you suggest to put this then in an absence of a way to provide an external class_hande callback?

Well I am also somewhat surprised about your arduino like approach. Regardless #ifdef or something else, it is wrong. You can not launch a bootloader from CDC ACM class implementation. Also, there will not be any callback for that in CDC ACM class implementation. The interface for the application is UART API. In the CDC ACM sample, you can find how to [obtain the baudrate, UART_LINE_CTRL_BAUD_RATE](https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/subsys/usb/cdc_acm/src/main.c#L121). If it can not work for you then there is no other way.

KubaFYI replied:
I know that doing it with the UART API would be cleaner abstraction-wise but because changes in DTR and the likes have to be polled manually, instead of having a callback, doing it this way would require a nasty spinning thread which only keeps polling the line.

> Also, there will not be any callback for that in CDC ACM class implementation.

There will not be any callback for what? The class changing? The DTR line? It's not clear the way you wrote it.

jfischer-phytec-iot replied:
>  know that doing it with the UART API would be cleaner abstraction-wise but because changes in DTR and the likes have to be polled manually, instead of having a callback, doing it this way would require a nasty spinning thread which only keeps polling the line.

You could use workqueue thread and timer for that. But it is your app and has nothing to do with CDC ACM implementation (or  subsys/power/reboot.c) or Zephyr RTOS self. And there are not _many_  boards but just few. What is the plan in general, to include USB device stack and ACM class per default? So if one would flash `samples/hello_world`, how to recover to bootloader? What you doing here is crap, stop it.



> There will not be any callback for what? The class changing? The DTR line? It's not clear the way you wrote it.

No.

KubaFYI replied:
Look, I don't really care if this is in Zephyr or not. Just thought it might be useful for other people who would like to make their boards retain Arduino-IDE compatibility. Your responses sound like you are either not understanding what I'm asking, or are being intentionally dismissive and generally unpleasant.

KubaFYI committed:
usb_cdc: sam0: Add an option to keep ardurno-style upgrade enter

Many of the Zephyr supported boards are Arduino compatible.
However after being flashed with Zephyr, even if they keep their
original bootloader they lose the ability to have the firmware
uploaded to them hands-free, from the Arduino IDE. This commit
allows for enabling of such ability.

Currently only enabled for adafruit_feather_m0_basic_proto board
but the code should work in principle with other adafruit M0 SoCs.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

mbolivar-nordic commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -51,14 +54,27 @@ def do_run(self, command, **kwargs):
+            sleep(0.250)
+            # Wait until port re-enumerates
+            retries = 0
+            while (retries < 3):
```
`while retries < 3:` since this is Python, not C

KubaFYI replied:
Ah yes. That was a silly one.

mbolivar-nordic commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -51,14 +54,27 @@ def do_run(self, command, **kwargs):
+                except CalledProcessError as e:
+                    if e.returncode == 1:
+                        retries += 1
+                        print("Retrying...")
```
Please use self.logger.info

KubaFYI replied:
Fixed.

mbolivar-nordic commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -51,14 +54,27 @@ def do_run(self, command, **kwargs):
+            retries = 0
+            while (retries < 3):
+                try:
+                    self.check_call(cmd_stty)
```
Am I imagining something, or did this used to just check if the file existed previously? Why is it calling stty again?

KubaFYI replied:
It was either there for checking if the file existed or for putting those ardurino-compatible boards in upgrade mode for flashing (though the latter might be only a coincidence). If the call to stty does indeed reboot a board into upgrade mode it causes it to briefly disappear from /dev/.  I included the following calls of stty in order to make sure that it re-appears by the time we try to move on to calling bossac.

mbolivar-nordic replied:
I don't know what this stty line is doing, but is it safe to call multiple times like this? It seems like just checking `os.path.isfile(self.port)` might work. Is there a reason to call stty instead of doing that?

Asking mostly out of concern about side effects. Any education you can provide me would be very welcome. Thank you.

mbolivar-nordic reviewed (commented):
Hmm, on second thought, I see some remaining nits and have one question.

KubaFYI committed:
testing: bossac: Make tests work with potential retries of the stty

The stty command is now potentially repeated as bossac port might
be re-enumerating. The tests need to be altered to reflect that.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

KubaFYI committed:
runners: bossac: Make sure the device re-enumerates after reboot

The runner opens the port at 1200 which is sometimes designed to
reboot a SoC into download mode. This causes the device port to
briefly disappear and so the runner needs to wait for it to
re-appear.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

hackwerken commented:
I was in the process of implementing this in an application. Unfortunately we are stuck with the Arduino bootloader because our product was originally based on an Arduino Zero (Atmel Samd21 Cortex M0+ SoC).

I implemented this by polling the uart_line_ctrl_get() function for baudrate and DTR. A callback would be nicer, but polling works as well. Anyway, it seems that the ACM driver often misses ether the baudrate or DTR command from the host (ie I need to run `stty --file "/dev/ttyACM0" 1200` many times to get it to trigger). @KubaFYI did you observe this as well? I'm asking because your bossac script does the retries. I have implemented this same (offal) bootloader trigger mechanism with the standard Atmel's libraries (ASF3) before. With ASF it never misses.

I didn't test this in isolation yet (without my other application code). I guess that is what I should try next.

KubaFYI commented:
Hi @hackwerken - no, I haven't observed the behaviour you mention but it sounds like a timing issue. Could it be due to your polling approach sometimes failing to detect the change in DTR? What's the interval between your polls and how does it compare to how frequently `cdc_acm_class_handle_req()` is called?

hackwerken commented:
@KubaFYI Thanks for your reply.

The polling currently doesn't happen at an deterministic frequency since or app is still single threaded. So it just polls 'as fast as possible'.

I didn't consider that the script just briefly sets the DTR bit and then closes again. So this seems impossible to implement with just polling. Even if I could poll faster, it will always kinda be a race condition waiting to happen.

A proper solution would be to add callbacks to the `uart_line_ctl` API. I found this issue #25613, but that suggests just adding it to the CDC API. In my opinion it should be part of the UART API, to keep it all consistent. This however would be a major effort, since it would have to be implemented in all UART drivers.

For now I think I will implement some other way of triggering the bootloader. Unfortunately this breaks backwards compatibility with our old Arduino based software. But I guess we can deal with that.

nzmichaelh commented on subsys/power/reboot.c:
```diff
@@ -33,3 +36,46 @@ void sys_reboot(int type)
+#endif
+
+#ifdef CONFIG_SOC_SERIES_SAMD21
+	while (!NVMCTRL->INTFLAG.reg & NVMCTRL_INTFLAG_READY) {
```
Instead of erasing the flash, you can put the DOUBLE_TAP_MAGIC value at the top of RAM.

nzmichaelh commented:
@mbolivar-nordic @jfischer-phytec-iot @KubaFYI see #26173 for another attempt at this.

Compared to this PR,

- It adds a 'line discipline changed' interface, keeping the bootloader code out of cdc_acm.c
- It adds a SAM0 specific handler, keeping the SAM0 code out of reboot.c
- It uses the RAM based stay-in-bootloader method instead of erasing flash

I've tested this on the Arduino Zero (with custom bootloader) and the Seeeduino XIAO, both with the Arduino and Adafruit UF2 bootloaders.

It does need a 'sleep' in bossa.py between the stty call and the bossa call to give the board time to reboot.

@KubaFYI WDYT about splitting out the bossa.py changes?

carlescufi commented:
Closing this since #26173 has been merged. @KubaFYI thanks for the contribution. Please open a new PR to follow-up on [this](https://github.com/zephyrproject-rtos/zephyr/pull/25422#issuecomment-643752690)
